paddle/fluid/operators/conv_mkldnn_op.cc

```diff
@@ -358,132 +364,424 @@ class ConvMKLDNNOpKernel : public paddle::framework::OpKernel<T> {
 
     std::vector<primitive> pipeline;
 
-    auto user_src_md = platform::MKLDNNMemDesc(
-        {src_tz}, platform::MKLDNNGetDataType<T>(), input->format());
-    auto user_weights_md = platform::MKLDNNMemDesc(
-        {weights_tz}, platform::MKLDNNGetDataType<T>(),
-        (g == 1) ? filter->format() : mkldnn::memory::format::goihw);
-
-    /* create memory descriptor for convolution without specified format
-     * ('any') which lets a primitive (convolution in this case) choose
-     * the memory format preferred for best performance
-     */
-    std::string data_format = ctx.Attr<std::string>("data_format");
-    auto chosen_memory_format =
-        platform::data_format_to_memory_format(data_format);
-
-    auto src_md = platform::MKLDNNMemDesc(
-        src_tz, platform::MKLDNNGetDataType<T>(), chosen_memory_format);
-    auto weights_md = platform::MKLDNNMemDesc(
-        weights_tz, platform::MKLDNNGetDataType<T>(), chosen_memory_format);
-    std::vector<int> bias_tz;  // TODO(mgallus): avoid empty vector creation.
-                               // Currently used whenever bias is != nullptr.
-    auto dst_md = platform::MKLDNNMemDesc(
-        dst_tz, platform::MKLDNNGetDataType<T>(), chosen_memory_format);
-
-    // create a conv primitive descriptor and save it for usage in backward
-    std::shared_ptr<mkldnn::convolution_forward::primitive_desc> conv_pd;
-    auto fwd_prop_kind = is_test ? mkldnn::prop_kind::forward_inference
-                                 : mkldnn::prop_kind::forward_training;
-    if (bias) {
-      bias_tz = paddle::framework::vectorize2int(bias->dims());
-      auto bias_md = platform::MKLDNNMemDesc(
-          bias_tz, platform::MKLDNNGetDataType<T>(), memory::format::x);
-      conv_pd = ConvFwdPrimitiveDesc(
-          src_md, weights_md, bias_md, dst_md, strides, paddings, mkldnn_engine,
-          fuse_relu, fuse_residual_conn, fwd_prop_kind);
-    } else {
-      conv_pd = ConvFwdPrimitiveDesc(src_md, weights_md, dst_md, strides,
-                                     paddings, mkldnn_engine, fuse_relu,
-                                     fuse_residual_conn, fwd_prop_kind);
-    }
-    // Save conv_pd/src_memory/weights_memory for backward pass
-    if (!is_test) dev_ctx.SetBlob(key_conv_pd, conv_pd);
-
-    ConvMKLDNNHandler handler(conv_pd, dev_ctx, mkldnn_engine, key);
+    bool is_INT8 = ctx.HasInput("Scale_in") ? true : false;
+    if (!is_INT8) {
+      auto user_src_md = platform::MKLDNNMemDesc(
```

Comment:
This if-else statement is very big and it's very difficult to follow. If-block is more than 100 lines of code. Could you move each part to separate functions?

Reply:
yes, we are planning to restructure the code on next PR, and keep these code for current PR, we want to commit it as soon as possible because calibration tool check also needs it.

Reply:
There are three code structures, and both of them have their own strengths and weaknesses.
1. Our current code is the first structure, it's easy for code merge and the logic is very clear but the if-else statement is big.
2. We can move each part to separate functions, but for reusing more code, we need to transfer many parameters to these functions like : input, filter, bias, output, src_tz, weights_tz, group ......
3. We can move each part to separate functions without reusing code.

I think the second way is better than others, what do you think?

Reply:
I also think the second option seems to be better.

Reply:
Regarding the second option: if you believe that there are too many parameters passed to the function, you can just pass the context instead and fetch necessary parameters from it.

Reply:
Synced with Poland team, we agreed to have such option in current implementation. @Sand3r- @tpatejko @kbinias